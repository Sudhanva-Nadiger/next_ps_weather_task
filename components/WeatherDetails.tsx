import { Detail, Weather } from '@/types';
import { Tabs, TabsContent, TabsList, TabsTrigger } from './ui/tabs';
import { absoluteUrl, getDifferenceBetweenDates, getWeatherUrl } from '@/lib/utils';
import { Globe2 } from 'lucide-react';
import Content from './Content';


// Revalidate data every 3 hours
const revalidate = 3600 * 3;

type Props = {
    q?: string
}

const WeatherDetails = async ({
    q
}: Props) => {
    if (!q || q === '') {
        return (
            <div className='flex flex-col items-center justify-center h-full'>
                <Globe2 className='w-16 h-16 text-muted-foreground' />
                <h1 className='text-3xl text-muted-foreground'>
                    Enter city name to get weather details.
                </h1>
            </div>
        )
    }

    const url = getWeatherUrl(q)

    const response = await fetch(url, {
        next: {
            revalidate: revalidate
        }
    })

    const data = await response.json() as Weather

    if(!data) {
        return (
            <div className='flex flex-col items-center justify-center h-full'>
                <Globe2 className='w-16 h-16 text-muted-foreground' />
                <h1 className='text-3xl text-muted-foreground'>
                    Something went wrong. Please try again later.
                </h1>
            </div>
        )
    }

    if(data.cod !== '200') {
        return (
            <div className='flex flex-col items-center justify-center h-full'>
                <Globe2 className='w-16 h-16 text-muted-foreground' />
                <h1 className='text-3xl text-muted-foreground'>
                    {data.message}
                </h1>
            </div>
        )
    } 
    
    

    const today = new Date()

    const weatherData: Array<Array<Detail>> = [[], [], []];

    data.list.forEach((item) => {
        const date = new Date(item.dt * 1000)

        const diff = getDifferenceBetweenDates(today, date)

        if(diff > 2) return;

        weatherData[diff].push(item)
    });

    return (
        <>
            <Tabs defaultValue="today" className='flex items-center flex-col'>
                <TabsList className='flex sm:flex-row flex-col justify-center gap-4 h-fit w-fit p-2 mt-6'>
                    <TabsTrigger className='shadow border w-[250px]' value="today">Today</TabsTrigger>
                    <TabsTrigger className='shadow border w-[250px]' value="tomorrow">Tomorrow</TabsTrigger>
                    <TabsTrigger className='shadow border w-[250px]' value="dayAfterTomorrow">Day After Tomorrow</TabsTrigger>
                </TabsList>
                <div className='w-full'>
                    <TabsContent value='today' className='w-full'>
                        <Content weatherData={weatherData[0]} title='Today' />
                    </TabsContent>
                    <TabsContent value='tomorrow' className='w-full'>
                        <Content weatherData={weatherData[1]} title='Tomorrow' />
                    </TabsContent>
                    <TabsContent value='dayAfterTomorrow' className='w-full'>
                        <Content weatherData={weatherData[2]} title='Day After Tomorrow' />
                    </TabsContent>
                </div>
            </Tabs>
        </>

    )
}

export default WeatherDetails