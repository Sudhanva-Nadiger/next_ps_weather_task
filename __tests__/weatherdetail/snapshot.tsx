import '@testing-library/jest-dom';
import '@testing-library/jest-dom/jest-globals';

import WeatherDetails from "@/components/WeatherDetails";
import { render } from "@testing-library/react";

const MOCK_DATA = {
    cod: '200',
    list: [
        {
            // todays date in unix timestamp
            dt: Math.floor(Date.now() / 1000),
            main: {
                temp: 20,
                feels_like: 20,
                pressure: 20,
                humidity: 20
            },
            weather: [
                {
                    icon: 'icon',
                    main: 'main',
                    description: 'description'
                }
            ],
            wind: {
                speed: 20
            }
        }
    ]
}

describe('Should match snapshot', () => {
    it('when no query is given', async () => {
        const jsx = await WeatherDetails({});
        const { container } = render(jsx);

        expect(container).toMatchSnapshot();
    })

    it("Should show the returned message if city name is incorrect", async () => {
        global.fetch = jest.fn(() => {
            return Promise.resolve({
                json: () => Promise.resolve({cod: '400', message: 'message'}),
              }) as any;
        });
        
        const jsx = await WeatherDetails({q: 'q'});
        const { container } = render(jsx);
        
        expect(container).toMatchSnapshot();
        
    });

    it("Should show tabs", async () => {
        global.fetch = jest.fn(() => {
            return Promise.resolve({
                json: () => Promise.resolve({cod: '200', list: []}),
              }) as any;
        });

        global.fetch = jest.fn(() => {
            return Promise.resolve({
                json: () => Promise.resolve({cod: '200', list: []}),
              }) as any;
        });
        
        const jsx = await WeatherDetails({q: 'q'});
        const { container } = render(jsx);

        expect(container).toMatchSnapshot();
    });

    it("Should show the details", async () => {
        global.fetch = jest.fn(() => {
            return Promise.resolve({
                json: () => Promise.resolve(MOCK_DATA),
              }) as any;
        });
        
        const jsx = await WeatherDetails({q: 'q'});
        const { container } = render(jsx);

        expect(container).toMatchSnapshot();
    });
})