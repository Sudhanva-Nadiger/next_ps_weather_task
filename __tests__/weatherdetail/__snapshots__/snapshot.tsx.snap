// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot Should show tabs 1`] = `
<div>
  <div
    class="flex items-center flex-col"
    data-orientation="horizontal"
    dir="ltr"
  >
    <div
      aria-orientation="horizontal"
      class="items-center rounded-md bg-muted text-muted-foreground flex sm:flex-row flex-col justify-center gap-4 h-fit w-fit p-2 mt-6"
      data-orientation="horizontal"
      role="tablist"
      style="outline: none;"
      tabindex="0"
    >
      <button
        aria-controls="radix-:r0:-content-today"
        aria-selected="true"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm shadow border w-[250px]"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="active"
        id="radix-:r0:-trigger-today"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Today
      </button>
      <button
        aria-controls="radix-:r0:-content-tomorrow"
        aria-selected="false"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm shadow border w-[250px]"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-:r0:-trigger-tomorrow"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tomorrow
      </button>
      <button
        aria-controls="radix-:r0:-content-dayAfterTomorrow"
        aria-selected="false"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm shadow border w-[250px]"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-:r0:-trigger-dayAfterTomorrow"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Day After Tomorrow
      </button>
    </div>
    <div
      class="w-full"
    >
      <div
        aria-labelledby="radix-:r0:-trigger-today"
        class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 w-full"
        data-orientation="horizontal"
        data-state="active"
        id="radix-:r0:-content-today"
        role="tabpanel"
        style="animation-duration: 0s;"
        tabindex="0"
      >
        <div
          class="flex flex-col items-center justify-center h-full"
        >
          <h1
            class="text-3xl text-muted-foreground"
          >
            No data available.
          </h1>
        </div>
      </div>
      <div
        aria-labelledby="radix-:r0:-trigger-tomorrow"
        class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 w-full"
        data-orientation="horizontal"
        data-state="inactive"
        hidden=""
        id="radix-:r0:-content-tomorrow"
        role="tabpanel"
        tabindex="0"
      />
      <div
        aria-labelledby="radix-:r0:-trigger-dayAfterTomorrow"
        class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 w-full"
        data-orientation="horizontal"
        data-state="inactive"
        hidden=""
        id="radix-:r0:-content-dayAfterTomorrow"
        role="tabpanel"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;

exports[`Should match snapshot Should show the details 1`] = `
<div>
  <div
    class="flex items-center flex-col"
    data-orientation="horizontal"
    dir="ltr"
  >
    <div
      aria-orientation="horizontal"
      class="items-center rounded-md bg-muted text-muted-foreground flex sm:flex-row flex-col justify-center gap-4 h-fit w-fit p-2 mt-6"
      data-orientation="horizontal"
      role="tablist"
      style="outline: none;"
      tabindex="0"
    >
      <button
        aria-controls="radix-:r4:-content-today"
        aria-selected="true"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm shadow border w-[250px]"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="active"
        id="radix-:r4:-trigger-today"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Today
      </button>
      <button
        aria-controls="radix-:r4:-content-tomorrow"
        aria-selected="false"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm shadow border w-[250px]"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-:r4:-trigger-tomorrow"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tomorrow
      </button>
      <button
        aria-controls="radix-:r4:-content-dayAfterTomorrow"
        aria-selected="false"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm shadow border w-[250px]"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-:r4:-trigger-dayAfterTomorrow"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Day After Tomorrow
      </button>
    </div>
    <div
      class="w-full"
    >
      <div
        aria-labelledby="radix-:r4:-trigger-today"
        class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 w-full"
        data-orientation="horizontal"
        data-state="active"
        id="radix-:r4:-content-today"
        role="tabpanel"
        style="animation-duration: 0s;"
        tabindex="0"
      >
        <div
          class="w-full h-full space-y-2 p-2"
        >
          <h1
            class="text-center text-2xl font-bold"
          >
            Today
             (
            4/20/2024
            )
          </h1>
          <div
            class="grid grid-cols-1 gap-4 h-full"
          >
            <div
              class="flex flex-col bg-secondary rounded-md p-4 gap-2"
            >
              <div
                class="flex gap-1 font-bold text-muted-foreground"
              >
                <p>
                  8:10:58 PM
                </p>
                -
                <p>
                  11:10:58 PM
                </p>
              </div>
              <div
                class="flex flex-col gap-2 sm:gap-1 sm:flex-row sm:max-h-52"
              >
                <div
                  class="flex flex-col min-w-[300px] sm:max-w-[300px] gap-1 items-center border rounded p-2"
                >
                  <img
                    alt="image"
                    class="w-[100px] aspect-square rounded-md"
                    data-nimg="1"
                    decoding="async"
                    height="50"
                    loading="lazy"
                    src="/_next/image?url=%2Ficons%2Ficon.png&w=128&q=75"
                    srcset="/_next/image?url=%2Ficons%2Ficon.png&w=64&q=75 1x, /_next/image?url=%2Ficons%2Ficon.png&w=128&q=75 2x"
                    style="color: transparent;"
                    width="50"
                  />
                  <h2
                    class="text-3xl font-extrabold"
                  >
                    20
                     Â°C
                  </h2>
                  <div
                    class="flex flex-col items-center gap-2 text-sm"
                  >
                    <h3
                      class="font-bold"
                    >
                      main
                    </h3>
                    <p>
                      description
                    </p>
                  </div>
                </div>
                <div
                  class="flex flex-col gap-1 min-w-[300px] sm:max-w-[300px] items-center border rounded px-3 py-2 justify-around"
                >
                  <h3
                    class="text-md font-[600] text-start w-full"
                  >
                    Details:
                  </h3>
                  <div
                    class="flex justify-between w-full"
                  >
                    <p>
                      <svg
                        class="lucide lucide-thermometer w-4 h-4 mr-2 inline-block"
                        fill="none"
                        height="24"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M14 4v10.54a4 4 0 1 1-4 0V4a2 2 0 0 1 4 0Z"
                        />
                      </svg>
                      Feels Like
                    </p>
                    <p
                      class="font-medium"
                    >
                      20
                       Â°C
                    </p>
                  </div>
                  <div
                    class="flex justify-between w-full"
                  >
                    <p>
                      <svg
                        class="lucide lucide-gauge w-4 h-4 mr-2 inline-block"
                        fill="none"
                        height="24"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="m12 14 4-4"
                        />
                        <path
                          d="M3.34 19a10 10 0 1 1 17.32 0"
                        />
                      </svg>
                      Pressure
                    </p>
                    <p
                      class="font-medium"
                    >
                      20
                       hPa
                    </p>
                  </div>
                  <div
                    class="flex justify-between w-full"
                  >
                    <p>
                      <svg
                        class="lucide lucide-wind w-4 h-4 mr-2 inline-block"
                        fill="none"
                        height="24"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M17.7 7.7a2.5 2.5 0 1 1 1.8 4.3H2"
                        />
                        <path
                          d="M9.6 4.6A2 2 0 1 1 11 8H2"
                        />
                        <path
                          d="M12.6 19.4A2 2 0 1 0 14 16H2"
                        />
                      </svg>
                      Wind
                    </p>
                    <p
                      class="font-medium"
                    >
                      20
                       m/s
                    </p>
                  </div>
                  <div
                    class="flex justify-between w-full"
                  >
                    <p>
                      <svg
                        class="lucide lucide-droplets w-4 h-4 mr-2 inline-block"
                        fill="none"
                        height="24"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M7 16.3c2.2 0 4-1.83 4-4.05 0-1.16-.57-2.26-1.71-3.19S7.29 6.75 7 5.3c-.29 1.45-1.14 2.84-2.29 3.76S3 11.1 3 12.25c0 2.22 1.8 4.05 4 4.05z"
                        />
                        <path
                          d="M12.56 6.6A10.97 10.97 0 0 0 14 3.02c.5 2.5 2 4.9 4 6.5s3 3.5 3 5.5a6.98 6.98 0 0 1-11.91 4.97"
                        />
                      </svg>
                      Humidity
                    </p>
                    <p
                      class="font-medium"
                    >
                      20
                       %
                    </p>
                  </div>
                </div>
                <div
                  class="relative overflow-hidden flex flex-col w-full gap-2 border rounded px-3 py-2"
                  dir="ltr"
                  style="position: relative; --radix-scroll-area-corner-width: 0px; --radix-scroll-area-corner-height: 0px;"
                >
                  <style>
                    [data-radix-scroll-area-viewport]{scrollbar-width:none;-ms-overflow-style:none;-webkit-overflow-scrolling:touch;}[data-radix-scroll-area-viewport]::-webkit-scrollbar{display:none}
                  </style>
                  <div
                    class="h-full w-full rounded-[inherit]"
                    data-radix-scroll-area-viewport=""
                    style="overflow-x: hidden; overflow-y: scroll;"
                  >
                    <div
                      style="min-width: 100%; display: table;"
                    >
                      <div
                        class="flex flex-col space-y-1"
                      >
                        <h2
                          class="font-semibold"
                        >
                          Message
                        </h2>
                        <hr
                          class="w-full border border-gray-300"
                        />
                      </div>
                      <ul
                        class="flex flex-col gap-2 list-disc ml-4 mt-2"
                      >
                        <li
                          class="font-normal text-lg"
                        >
                          It is windy outside. Watch out! ð¨.
                        </li>
                      </ul>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-labelledby="radix-:r4:-trigger-tomorrow"
        class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 w-full"
        data-orientation="horizontal"
        data-state="inactive"
        hidden=""
        id="radix-:r4:-content-tomorrow"
        role="tabpanel"
        tabindex="0"
      />
      <div
        aria-labelledby="radix-:r4:-trigger-dayAfterTomorrow"
        class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 w-full"
        data-orientation="horizontal"
        data-state="inactive"
        hidden=""
        id="radix-:r4:-content-dayAfterTomorrow"
        role="tabpanel"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;

exports[`Should match snapshot Should show the returned message if city name is incorrect 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center h-full"
  >
    <svg
      class="lucide lucide-earth w-16 h-16 text-muted-foreground"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21.54 15H17a2 2 0 0 0-2 2v4.54"
      />
      <path
        d="M7 3.34V5a3 3 0 0 0 3 3v0a2 2 0 0 1 2 2v0c0 1.1.9 2 2 2v0a2 2 0 0 0 2-2v0c0-1.1.9-2 2-2h3.17"
      />
      <path
        d="M11 21.95V18a2 2 0 0 0-2-2v0a2 2 0 0 1-2-2v-1a2 2 0 0 0-2-2H2.05"
      />
      <circle
        cx="12"
        cy="12"
        r="10"
      />
    </svg>
    <h1
      class="text-3xl text-muted-foreground"
    >
      message
    </h1>
  </div>
</div>
`;

exports[`Should match snapshot when no query is given 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center h-full"
  >
    <svg
      class="lucide lucide-earth w-16 h-16 text-muted-foreground"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21.54 15H17a2 2 0 0 0-2 2v4.54"
      />
      <path
        d="M7 3.34V5a3 3 0 0 0 3 3v0a2 2 0 0 1 2 2v0c0 1.1.9 2 2 2v0a2 2 0 0 0 2-2v0c0-1.1.9-2 2-2h3.17"
      />
      <path
        d="M11 21.95V18a2 2 0 0 0-2-2v0a2 2 0 0 1-2-2v-1a2 2 0 0 0-2-2H2.05"
      />
      <circle
        cx="12"
        cy="12"
        r="10"
      />
    </svg>
    <h1
      class="text-3xl text-muted-foreground"
    >
      Enter city name to get weather details.
    </h1>
  </div>
</div>
`;
